### PR TITLE
PT5353 Adding collection results to processing Job

### PR DESCRIPTION
**Pull request by andher1802**
(insert scope and description here)

Items:
* [x] Implemented (new) unit tests
* [ ] Updated [documentation](sdk.up42.com)

For release:
* [x] Bumped version
* [x] Added changelog

### CODE REVIEW

File: up42/processing.py [hunk 1]
```diff
@@ -53,20 +54,34 @@ class Job:
...
+    collection: Optional[pystac.Collection] = None
     started: Optional[datetime.datetime] = None
     finished: Optional[datetime.datetime] = None
 
     @staticmethod
     def __to_datetime(value: Optional[str]):
         return value and datetime.datetime.fromisoformat(value.rstrip("Z"))
 
+    @staticmethod
+    def _pystac_client():
```
Comment (javidq) [inline up42/processing.py:66]: Let's add StacClient descriptor to base instead

File: up42/processing.py [hunk 2]
```diff
@@ -35,6 +35,7 @@ class JobMetadata(TypedDict):
     accountID: str  # pylint: disable=invalid-name
     workspaceID: Optional[str]  # pylint: disable=invalid-name
     definition: dict
+    results: dict
```
Comment (javidq) [inline up42/processing.py:38]: Let's add typed dict for its contents

File: up42/processing.py [hunk 3]
```diff
@@ -53,20 +54,34 @@ class Job:
...
+                collection_id = metadata["results"]["collection"].split("/")[-1]
+                return Job._pystac_client().get_collection(collection_id)
+            return None
+
         return Job(
             process_id=metadata["processID"],
             id=metadata["jobID"],
             account_id=metadata["accountID"],
             workspace_id=metadata["workspaceID"],
+            collection=get_metadata_collection(metadata),
```
Comment (javidq) [inline up42/processing.py:84]: collection_url =  metadata.get("results", {}).get("collection")

Comment (javidq) [reply up42/processing.py:84]: and 
```python
@property
def collection(self):
# retrieve using self.stac_client and self.collection_url
```

**Review by javidq — CHANGES_REQUESTED** [review]

File: up42/processing.py [hunk 4]
```diff
@@ -64,16 +70,26 @@ class Job:
...
     started: Optional[datetime.datetime] = None
     finished: Optional[datetime.datetime] = None
 
+    @property
+    def collection(self) -> Optional[pystac.Collection]:
+        if self.collection_url is None:
+            return None
+        collection_id = self.collection_url.split("/")[-1]
+        client: base.StacClient = base.StacClient()
+        return client.get_collection(collection_id)  # type: ignore
```
Comment (andher1802) [inline up42/processing.py:83]: type: ignore, required due to static mypy analysis is not able to infer the dynamic properties of a descriptor.

File: up42/processing.py [hunk 5]
```diff
@@ -29,12 +30,18 @@ class JobStatus(enum.Enum):
     RELEASED = "released"
 
 
+class JobResults(TypedDict):
+    collection: str
```
Comment (javidq) [inline up42/processing.py:34]: Optional[str]

File: up42/processing.py [hunk 6]
```diff
@@ -64,16 +71,29 @@ class Job:
     status: JobStatus
     created: datetime.datetime
     updated: datetime.datetime
+    collection_url: Optional[str] = None
     started: Optional[datetime.datetime] = None
     finished: Optional[datetime.datetime] = None
 
+    @property
+    def _client(self) -> pystac_client.Client:
+        return base.StacClient()  # type: ignore
```
[the remark this answers is not in the document]
Comment (javidq) [reply up42/processing.py:80]: Let's add instead
```python
stac_client = base.StacClient()
```
after session

File: up42/processing.py [hunk 7]
```diff
@@ -29,12 +30,18 @@ class JobStatus(enum.Enum):
     RELEASED = "released"
 
 
+class JobResults(TypedDict):
+    collection: str
+    # To do: include errors
```
Comment (javidq) [inline up42/processing.py:35]: `# TODO: include errors` - otherwise this is not detected by IDEs

File: tests/test_processing.py [hunk 8]
```diff
@@ -282,9 +316,19 @@ def test_should_provide_inputs(self, requests_mock: req_mock.Mocker):
...
+    @mock.patch("up42.base.StacClient")
+    def test_should_get_job(self, mock_stac_client, requests_mock: req_mock.Mocker):
+        mock_pystac_client = mock.MagicMock()
+        mock_pystac_client.get_collection.return_value = COLLECTION
+        mock_stac_client.return_value = mock_pystac_client
         requests_mock.get(url=JOB_URL, json=JOB_METADATA)
         assert processing.Job.get(JOB_ID) == JOB
+        collection = JOB.collection
+        assert collection == COLLECTION
+        assert collection.id == COLLECTION_ID
```
Comment (javidq) [inline tests/test_processing.py:328]: To be moved to a dedicated test `test_should_get_collection`

File: tests/test_processing.py [hunk 9]
```diff
@@ -282,9 +316,19 @@ def test_should_provide_inputs(self, requests_mock: req_mock.Mocker):
...
+
+    def test_collection_property_with_none_url(self):
```
Comment (javidq) [inline tests/test_processing.py:330]: `test_should_get_no_collection_if_collection_url_is_missing`

File: tests/test_processing.py [hunk 10]
```diff
@@ -282,9 +316,19 @@ def test_should_provide_inputs(self, requests_mock: req_mock.Mocker):
...
+        assert JOB_NO_COLLECTION_URL.collection is None
```
Comment (javidq) [inline tests/test_processing.py:331]:
```python
job = dataclasses.replace(JOB, collection_url=None)
assert not job.collection
```

File: tests/test_processing.py [hunk 11]
```diff
@@ -197,7 +227,11 @@ def test_should_construct(self, requests_mock: req_mock.Mocker, cost: processing
         assert not template.errors
         assert template.cost == cost
 
-    def test_should_execute(self, requests_mock: req_mock.Mocker):
+    @mock.patch("up42.base.StacClient")
+    def test_should_execute(self, mock_stac_client, requests_mock: req_mock.Mocker):
+        mock_pystac_client = mock.MagicMock()
+        mock_pystac_client.get_collection.return_value = COLLECTION
+        mock_stac_client.return_value = mock_pystac_client
```
Comment (javidq) [inline tests/test_processing.py:234]: These changes are not needed

File: tests/test_processing.py [hunk 12]
```diff
@@ -61,6 +79,18 @@
...
+    account_id=ACCOUNT_ID,
+    workspace_id=constants.WORKSPACE_ID,
+    definition=DEFINITION,
+    collection_url=COLLECTION_URL,
+    status=processing.JobStatus.CREATED,
+    created=NOW,
+    updated=NOW,
+)
+
+JOB_NO_COLLECTION_URL = processing.Job(
```
Comment (javidq) [inline tests/test_processing.py:93]: not needed

File: tests/test_processing.py [hunk 13]
```diff
@@ -20,7 +20,24 @@
 COST_URL = f"{constants.API_HOST}/v2/processing/processes/{PROCESS_ID}/cost"
 EXECUTION_URL = f"{constants.API_HOST}/v2/processing/processes/{PROCESS_ID}/execution"
 TITLE = "title"
-ITEM_URL = "https://item-url"
+COLLECTION_ID = str(uuid.uuid4())
+ITEM_ID = uuid.uuid4()
```
Comment (javidq) [inline tests/test_processing.py:24]: Not used and not needed

File: tests/test_processing.py [hunk 14]
```diff
@@ -20,7 +20,24 @@
...
+COLLECTION_URL = f"https://collection-url/v2/assets/stac/collections/{COLLECTION_ID}"
```
Comment (javidq) [inline tests/test_processing.py:25]: `f"https://collections/{COLLECTION_ID}"`

File: tests/test_processing.py [hunk 15]
```diff
@@ -20,7 +20,24 @@
...
+ITEM_URL = "https://item-url/"
+COLLECTION = pystac.Collection.from_dict(
```
Comment (javidq) [inline tests/test_processing.py:27]: Let's rather have `collection = mock.MagicMock()` in the only test which uses it

File: tests/test_base.py [hunk 16]
```diff
@@ -69,3 +74,14 @@ def test_workspace_id_should_provide_default_value(self):
         assert record.workspace_id == constants.WORKSPACE_ID
         assert ActiveRecord.class_workspace_id == constants.WORKSPACE_ID
         assert constants.WORKSPACE_ID in repr(record)
+
+    @mock.patch.object(utils, "stac_client")
```
Comment (javidq) [inline tests/test_base.py:78]: We should auth object above, but not pystac client and utils

**Review by javidq — CHANGES_REQUESTED** [review]

**Review by javidq — APPROVED** [review]

**Pull request merged**